### PR TITLE
[Product] added availableTo parameter

### PR DESCRIPTION
**Pull request by starspire**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Sometimes except availableOn product parameter we need also availableTo one. This is especially usefull when You selling something like concert or training tickets which events take place at particular day and it is useless to selling them after that day. When You leave this parameter blank nothing new happens but when You set a date, Sylius will validate it with availableOn in isAvailable method.

This is my first bigger commit to Sylius so any tips how to improve my code are highly appreciated. Hope it will be usefull for community.

### CODE REVIEW

File: src/Sylius/Component/Product/Model/Product.php [hunk 1]
```diff
@@ -189,7 +194,7 @@ public function setMetaDescription($metaDescription)
      */
     public function isAvailable()
     {
-        return new \DateTime() >= $this->availableOn;
+        return new \DateTime() >= $this->availableOn && (!$this->availableTo || $this->availableTo >= new \DateTime());
```
Comment (tristanbes) [inline src/Sylius/Component/Product/Model/Product.php]: you should invert your comparaison order and make it like: 
```
new \DateTime() <= $this->availableTo;
```

to keep the same logic with the check before

File: src/Sylius/Component/Product/Model/Variant.php [hunk 2]
```diff
@@ -52,7 +57,7 @@ public function setProduct(ProductInterface $product = null)
      */
     public function isAvailable()
     {
-        return new \DateTime() >= $this->availableOn;
+        return new \DateTime() >= $this->availableOn && (!$this->availableTo || $this->availableTo >= new \DateTime());
```
Comment (tristanbes) [inline src/Sylius/Component/Product/Model/Variant.php]: same remark as before

**Comment by pamil** [comment]
It would be great to have some `DateRange` VO that can tell us if date is between start and end date.

**Comment by starspire** [comment]
@pamil Did you mean a simple class called DateRange, with 2 parameters in constructor and boolean methods like isInRange() ?

**Comment by pamil** [comment]
Something like that, another idea is to not require these dates, as no date would mean no restriction. So empty `DateRange` would match every other date, `DateRange` with start date only would match every date past the start date etc.

It's not essential to do it in this PR, but I think it would be a nice addition and it would make our codebase cleaner (especially when we can use Doctrine ORM 2.5 with embeddables support :)).

File: src/Sylius/Component/Product/Model/Variant.php (hunk 2 above)
Comment (starspire) [inline src/Sylius/Component/Product/Model/Variant.php]: @tristanbes Fixed.

**Comment by starspire** [comment]
@pamil I can write it Today, just tell me when I should to put this class, because I'm not sure now.

**Comment by aramalipoor** [comment]
Maybe `availableUntil` is more convenient here, I thought you're PR is about restricting a product **to** some users or groups ;)

**Comment by starspire** [comment]
@aramalipoor You are right, I'll change it.

**Comment by pamil** [comment]
@starspire as we don't have any common package shared in all other Sylius packages, let's put it in Product component and bundle (if needed), it'd be easy to move it somewhere else if there is a need for it.

**Comment by starspire** [comment]
@pamil DateRange class added. I think the whole work is done now :)

File: app/migrations/Version20151221151112.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() != 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('DROP INDEX fulltext_search_idx ON sylius_search_index');
+        $this->addSql('CREATE INDEX fulltext_search_idx ON sylius_search_index (item_id)');
+        $this->addSql('ALTER TABLE sylius_product_variant CHANGE available_to available_until DATETIME DEFAULT NULL');
```
Comment (aramalipoor) [inline app/migrations/Version20151221151112.php]: I think you don't need this migration when you have the other one. Just change `available_to` to `available_util` in previous migration.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<?php
+/**
+ * User: Krzysztof Wędrowicz krzysztof@wedrowicz.me
+ * Date: 21.12.15
+ * Time: 15:41
+ */
```
Comment (aramalipoor) [inline src/Sylius/Component/Product/Model/DateRange.php]: Fix copyright, do it like other files ;)

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+     * @var \DateTime
+     */
+    protected $availableOn;
+
+    /**
+     * @var \DateTime
+     */
+    protected $availableUntil;
+
+
```
Comment (aramalipoor) [inline src/Sylius/Component/Product/Model/DateRange.php]: Extra space.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    public function __construct($availableOn = null, $availableUntil = null)
+    {
+        $this->availableOn = $availableOn;
+        $this->availableUntil = $availableUntil;
+    }
+
+    public function isInRange()
+    {
+        $now = new \DateTime();
+        if($this->availableOn && $this->availableUntil)
```
Comment (aramalipoor) [inline src/Sylius/Component/Product/Model/DateRange.php]: Use `{` and `}`

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+     */
+    protected $availableUntil;
+
+
+    public function __construct($availableOn = null, $availableUntil = null)
+    {
+        $this->availableOn = $availableOn;
+        $this->availableUntil = $availableUntil;
+    }
+
```
Comment (aramalipoor) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing doc here for `isInRange`

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+namespace Sylius\Component\Product\Model;
+
+class DateRange
+{
+    /**
+     * @var \DateTime
+     */
+    protected $availableOn;
```
Comment (aramalipoor) [inline src/Sylius/Component/Product/Model/DateRange.php]: I think it's better to name this `$start`

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * @var \DateTime
+     */
+    protected $availableUntil;
```
Comment (aramalipoor) [inline src/Sylius/Component/Product/Model/DateRange.php]: I think it's better to name this `$end`

File: src/Sylius/Component/Product/Model/VariantInterface.php [hunk 10]
```diff
@@ -42,4 +42,14 @@ public function getAvailableOn();
...
+
+    /**
+     * @return \DateTime
+     */
+    public function getAvailableUntil();
+
+    /**
+     * @param null|\DateTime $availableUntil
+     */
+    public function setAvailableUntil(\DateTime $availableUntil = null);
```
Comment (aramalipoor) [inline src/Sylius/Component/Product/Model/VariantInterface.php]: I think it's not right to have default value of null here!? When are you calling it with no arguments? If you want to set to null you just `set(null)`

**Comment by aramalipoor** [comment]
@pamil Is it possible to have a `DateRange` kind of thing in only **one** field? That would be a useful thing if possbile ;)

File: src/Sylius/Component/Product/Model/VariantInterface.php (hunk 10 above)
Comment (starspire) [inline src/Sylius/Component/Product/Model/VariantInterface.php]: I just followed the convention above. There was already setAvailableOn with default null value.

**Comment by starspire** [comment]
@aramalipoor Thanks for your tips, the code was corrected.

**Comment by starspire** [comment]
@pjedrzejewski Could you take a look if this PR is ready to merged? I think it has been done.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 11]
```diff
@@ -0,0 +1,47 @@
+<?php
+
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing license block.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * @author Krzysztof Wędrowicz <krzysztof@wedrowicz.me>
+ */
+class DateRange
+{
+    /**
+     * @var \DateTime
+     */
+    protected $start;
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Product/Model/DateRange.php]: Should be private.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * @var \DateTime
+     */
+    protected $end;
+    
+    public function __construct($start, $end)
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing docblock.

Comment (pjedrzejewski) [inline src/Sylius/Component/Product/Model/DateRange.php]: And typehints.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    /**
+     * Checks whether current datetime is between start and end.
+     *
+     * @return bool
+     */
+    public function isInRange()
+    {
+        $now = new \DateTime();
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Product/Model/DateRange.php]: CS and I believe this could be a bit cleaner.

**Comment by pjedrzejewski** [comment]
Is there really no core php class that we could use? I don't think Product component is the right place for such generic class.

**Comment by starspire** [comment]
@pjedrzejewski I think there isn't. At least php.net (http://php.net/manual/en/book.datetime.php) doesn't know anything about it. DatePeriod is also something different from what we want. Eventually I found http://period.thephpleague.com/ but also this isn't strictly what we needed (but it has boolean contains() and overlaps() methods). I agree that current location isn't the most fortune one, but was suggested by @pamil to me, as the best of weak choices. I'm with pleasure make this in the way you decide but must to know how :)

**Comment by starspire** [comment]
@pjedrzejewski How about now? I followed your notices but nothings done in this PR for a week and because I already need this in my project I'd like to ask if I can still improve something or it is good enough to merge.

**Comment by pjedrzejewski** [comment]
@starspire That's a bit late from my side, but I did not notice that there are no specs. Would be cool if you can add them, do you know how to write them?

**Comment by starspire** [comment]
@pjedrzejewski I'm not mastering it, but think that can write some looking on examples of others. Will be in a few hours :)

**Comment by starspire** [comment]
@pjedrzejewski Done. Could you see why Travis throw Segmentation fault in PHP7?

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * @var \DateTime
+     */
+    private $end;
+
+    /**
+     * @param \DateTime $start
+     * @param \DateTime $end
+     */
+    public function __construct($start, $end)
```
Comment (lchrusciel) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing typehint.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * @return \DateTime
+     */
+    public function getEnd()
+    {
+        return $this->end;
+    }
+}
```
Comment (tristanbes) [inline src/Sylius/Component/Product/Model/DateRange.php]: missing EOF new lien

Comment (starspire) [inline src/Sylius/Component/Product/Model/DateRange.php]: Added.

**Comment by starspire** [comment]
@pjedrzejewski @pamil @aramalipoor @tristanbes @lchrusciel All of your suggestions were used :) Is there anything else to improve?

**Comment by pjedrzejewski** [comment]
@starspire Nice, it is good to go now but I have to ask you for last little thing. :) Could you please squash your commits? 14 is a bit too much for such small change. Please squash it to a single commit.

https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

Thanks!

**Comment by starspire** [comment]
@pjedrzejewski Done. Everything passed :)

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 17]
```diff
@@ -0,0 +1,86 @@
+<?php
+/*
```
Comment (stloyd) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing new line before this comment.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Product\Model;
+
+/**
+ * @author Krzysztof Wędrowicz <krzysztof@wedrowicz.me>
+ */
+class DateRange
```
Comment (stloyd) [inline src/Sylius/Component/Product/Model/DateRange.php]: IMO this class should be `final`.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * @var \DateTime
+     */
+    private $end;
+
+    /**
+     * @param \DateTime $start
+     * @param \DateTime $end
+     */
+    public function __construct(\DateTime $start = null,\DateTime $end = null)
```
Comment (stloyd) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing space after `,`.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+     * Checks whether current datetime meets two requirements:
+     * is greater than $start or equal,
+     * is smaller than $end or equal.
+     * Checks only these conditions which values ($start, $end) are not null.
+     *
+     * @return bool
+     */
+    public function isInRange()
+    {
+        if($this->start > new \DateTime() || ($this->end !== null && $this->end < new \DateTime())){
```
Comment (stloyd) [inline src/Sylius/Component/Product/Model/DateRange.php]: This should be splitted, and this is wrong as `start` also can be `null`.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @param \DateTime $start
+     */
+    public function setStart($start)
```
Comment (stloyd) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing typehint.

File: src/Sylius/Component/Product/Model/DateRange.php [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        $this->start = $start;
+    }
+
+    /**
+     * @param \DateTime $end
+     */
+    public function setEnd($end)
```
Comment (stloyd) [inline src/Sylius/Component/Product/Model/DateRange.php]: Missing typehint.

File: src/Sylius/Component/Product/Model/ProductInterface.php [hunk 23]
```diff
@@ -41,4 +41,16 @@ public function getAvailableOn();
...
+     * @return \DateTime
+     */
+    public function getAvailableUntil();
+
+    /**
+     * @param null|\DateTime $availableUntil
+     */
+    public function setAvailableUntil(\DateTime $availableUntil = null);
+
+
```
Comment (stloyd) [inline src/Sylius/Component/Product/Model/ProductInterface.php]: Those two new lines should be removed.

File: src/Sylius/Component/Product/Model/DateRange.php (hunk 20 above)
Comment (starspire) [inline src/Sylius/Component/Product/Model/DateRange.php]: Nope, because null cant be greater than anything so this condition would be redundant :)

**Comment by pjedrzejewski** [comment]
Thanks Krzysztof! :+1:

File: src/Sylius/Component/Product/Model/DateRange.php (hunk 20 above)
Comment (lchrusciel) [inline src/Sylius/Component/Product/Model/DateRange.php]: Anyway we used to use yoda style comparison: ``null !== $this->end``

File: src/Sylius/Component/Product/Model/DateRange.php (hunk 16 above)
Comment (lchrusciel) [inline src/Sylius/Component/Product/Model/DateRange.php]: Too many blank lines :P

File: src/Sylius/Component/Product/Model/DateRange.php (hunk 20 above)
Comment (tristanbes) [inline src/Sylius/Component/Product/Model/DateRange.php]: missing whitespace right after `ìf`

**Pull request merged**